### PR TITLE
Remove "public" tenant

### PR DESCRIPTION
**Pull request by rbradford**
Copy the methodology used for images onto workloads allowing us to remove the pseudo-tenant "public". Workloads gain a visibility field which is persisted into the database and the cache is modified to have slices of IDs for workloads associated with a given tenant or which are public.

Fixes: #1328

### CODE REVIEW

File: ciao-controller/api/api_test.go [hunk 1]
```diff
@@ -149,15 +149,15 @@ var tests = []test{
 		"",
 		fmt.Sprintf("application/%s", WorkloadsV1),
 		http.StatusOK,
-		`{"id":"ba58f471-0735-4773-9550-188e2d012941","description":"testWorkload","fw_type":"legacy","vm_type":"qemu","image_name":"","config":"this will totally work!","defaults":null,"storage":null}`,
+		`{"id":"ba58f471-0735-4773-9550-188e2d012941","description":"testWorkload","fw_type":"legacy","vm_type":"qemu","image_name":"","config":"this will totally work!","defaults":null,"storage":null,"visibility":""}`,
```
Comment (markdryan) [inline ciao-controller/api/api_test.go]: Is visibility = "" a valid value?

File: ciao-controller/internal/datastore/sqlite3db.go [hunk 2]
```diff
@@ -872,47 +862,39 @@ func (ds *sqliteDB) updateWorkload(w types.Workload) error {
...
+	for _, d := range w.Defaults {
+		err := ds.createWorkloadDefault(tx, w.ID, d)
+		if err != nil {
+			_ = tx.Rollback()
+			return err
+		}
+	}
+
+	// add in any workload storage resources
+	if len(w.Storage) > 0 {
```
Comment (markdryan) [inline ciao-controller/internal/datastore/sqlite3db.go]: This test is not needed.

File: ciao-controller/internal/datastore/sqlite3db.go [hunk 3]
```diff
@@ -872,47 +862,39 @@ func (ds *sqliteDB) updateWorkload(w types.Workload) error {
...
-		filename := fmt.Sprintf("%s_config.yaml", w.ID)
-		path := fmt.Sprintf("%s/%s", ds.workloadsPath, filename)
-		err := ioutil.WriteFile(path, []byte(w.Config), 0644)
-		if err != nil {
-			_ = tx.Rollback()
-			return err
-		}
+	// write config to file.
+	filename := fmt.Sprintf("%s_config.yaml", w.ID)
+	path := fmt.Sprintf("%s/%s", ds.workloadsPath, filename)
```
Comment (markdryan) [inline ciao-controller/internal/datastore/sqlite3db.go]: filepath.Join would be better here.

File: ciao-controller/internal/datastore/datastore.go [hunk 4]
```diff
@@ -245,6 +250,34 @@ func (ds *Datastore) initImages() error {
...
+	if err != nil {
+		return errors.Wrap(err, "error getting workloads from database")
+	}
+
+	for _, wl := range workloads {
+		ds.workloads[wl.ID] = wl
+
+		if wl.Visibility == types.Public {
+			ds.publicWorkloads = append(ds.publicWorkloads, wl.ID)
+		}
```
Comment (markdryan) [inline ciao-controller/internal/datastore/datastore.go]: Can workloads be public and belong to a tenant?

File: ciao-controller/internal/datastore/sqlite3db_test.go [hunk 5]
```diff
@@ -981,17 +981,18 @@ users:
...
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	if len(tenant.workloads) != 1 {
-		t.Fatal("Expected a workload associated with tenant")
+	var wl2 types.Workload
+	for i, w := range workloads {
+		if w.ID == wl.ID {
+			wl2 = workloads[i]
```
Comment (markdryan) [inline ciao-controller/internal/datastore/sqlite3db_test.go]: Do we need a break here?

File: ciao-controller/internal/datastore/datastore.go (hunk 4 above)
Comment (rbradford) [reply ciao-controller/internal/datastore/datastore.go]: Inconsistently. I've fixed it so that it they're exclusive.

**Comment by rbradford** [comment]
@markdryan integrated all your feedback, thanks (and fixed optional nits too.)

**Pull request merged**